src/shamir.rs

```diff
@@ -0,0 +1,419 @@
+//!
+//! This module contains the implementation of [`Shamir secret sharing`] using finite fields.
+//!
+//! [`Shamir secret sharing`](https://en.wikipedia.org/wiki/Shamir%27s_Secret_Sharing)
+//!
+use crate::field::{Field, Fp31};
+use rand::Rng;
+use rand_core::RngCore;
+use std::iter::repeat_with;
+use std::num::NonZeroU8;
+use std::ops::Add;
+use thiserror::Error;
+
+/// Interpolating polynomial used to reconstruct secrets encoded with Shamir secret scheme.
+#[derive(Debug)]
+pub struct LagrangePolynomial<F> {
+    coefficients: Vec<F>,
+}
+
+impl<F: Field> LagrangePolynomial<F> {
+    /// Constructs new polynomial of a given degree.
+    ///
+    /// ## Errors
+    /// Returns an error if it fails to evaluate at least one of the coefficients
+    pub fn new(degree: NonZeroU8) -> Result<Self, Error> {
+        if u128::from(degree.get()) >= F::PRIME.into() {
+            // SAFETY: F::Prime <= u8::MAX here
+            #[allow(clippy::cast_possible_truncation)]
+            Err(Error::FieldSizeError {
+                field_size: F::PRIME.into() as u8,
+                polynomial_degree: degree.get(),
+            })
+        } else {
+            let mut coefficients = Vec::with_capacity(degree.get() as usize);
+            let n = degree.get();
+            for i in 1..=n {
+                let mut x = F::ONE;
+                let mut denom = F::ONE;
+
+                let x_i = F::from(u128::from(i));
+                for j in 1..=n {
+                    if i != j {
+                        let x_j = F::from(u128::from(j));
+                        x *= x_j;
+                        denom *= x_j - x_i;
+                    }
+                }
+
+                coefficients.push(x * denom.invert());
+            }
+
+            Ok(Self { coefficients })
+        }
+    }
+
+    #[must_use]
+    #[allow(clippy::cast_possible_truncation)]
+    pub fn degree(&self) -> u8 {
+        // SAFETY: new() does not allow constructing more than u8::MAX coefficients
+        self.coefficients.len() as u8
+    }
+
+    // Evaluates P(0) using the first N points provided, ignoring extra points, if any.
+    fn evaluate_at_zero<T: Clone + Into<F>>(&self, points: &[T]) -> F {
+        points
+            .iter()
+            .zip(self.coefficients.iter())
+            .map(|(y, &x)| y.clone().into() * x)
+            .fold(F::ZERO, |lhs, rhs| lhs + rhs)
+    }
```

Comment:
This is a bit tricky. The order of the points you provide must be ordered the same as the order of the helpers.

I suggest adding a comment about that maybe?